### PR TITLE
hybrid retrieval

### PR DESCRIPTION
**Pull request by hbertrand**

### CODE REVIEW

File: buster/retriever/base.py [hunk 1]
```diff
@@ -18,10 +17,19 @@
...
+        thresh: float,
+        embedding_fn: Callable[[str], np.ndarray],
+        sparse_embedding_fn: Callable[[str], dict[str, list[float]]] = None,
+        *args,
+        **kwargs,
+    ):
         self.top_k = top_k
         self.thresh = thresh
-        self.embedding_model = embedding_model
+        self.get_embedding = lru_cache(embedding_fn)
```
Comment (jerpint) [inline buster/retriever/base.py:31]: was lru cache still needed?

Comment (hbertrand) [reply buster/retriever/base.py]: no, removed

File: buster/documents_manager/base.py [hunk 2]
```diff
@@ -58,7 +57,8 @@ def add(
         self,
         df: pd.DataFrame,
         num_workers: int = 16,
-        embedding_fn: callable = get_openai_embedding,
+        embedding_fn: Callable[[str], np.ndarray] = get_openai_embedding,
+        sparse_embedding_fn: Callable[[str], dict[str, list[float]]] = None,
```
Comment (jerpint) [inline buster/documents_manager/base.py:61]: can you add docstrings for what sparse_embeddinf_fn does and when it shoudl be used?

File: buster/documents_manager/service.py [hunk 3]
```diff
@@ -45,6 +45,8 @@ def get_source_id(self, source: str) -> str:
     def _add_documents(self, df: pd.DataFrame):
         """Write all documents from the dataframe into the db as a new version."""
 
+        use_sparse_vector = "sparse_embedding" in df.columns
```
Comment (jerpint) [inline buster/documents_manager/service.py:48]: add a log message here saying its enable/disabled?

File: buster/documents_manager/service.py [hunk 4]
```diff
@@ -53,14 +55,28 @@ def _add_documents(self, df: pd.DataFrame):
...
 
                 document_id = str(self.db.documents.insert_one(document).inserted_id)
-                self.index.upsert([(document_id, embedding, {"source": source})], namespace=self.namespace)
+                vector = {"id": document_id, "values": embedding, "metadata": {"source": source}}
+                if use_sparse_vector:
+                    vector["sparse_values"] = sparse_embedding
+
+                to_upsert.append(vector)
+
+            MAX_PINECONE_BATCH_SIZE = 100 if use_sparse_vector else 1000
```
Comment (jerpint) [inline buster/documents_manager/service.py:77]: what is this batch size and why is it important it be different? maybe add a comment explaining the nuance?

File: buster/busterbot.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ class BusterConfig:
             "max_tokens": 3000,
             "top_k": 3,
             "thresh": 0.7,
-            "embedding_model": "text-embedding-ada-002",
+            "embedding_fn": get_openai_embedding,
```
[the remark this answers is not in the document]
Comment (hbertrand) [reply buster/busterbot.py:36]: yes it's rebased. Not sure why it still shows this

[the remark this answers is not in the document]
Comment (hbertrand) [reply buster/busterbot.py:36]: wait I checked main, we didn't do the change actually. Seems like a bug fix then :-)

**Review by jerpint — APPROVED** [review]

**Pull request merged**